diff --git a/plugins/thunar-uca/thunar-uca-model.c b/plugins/thunar-uca/thunar-uca-model.c
index d5cb8e4..808ed49 100644
--- a/plugins/thunar-uca/thunar-uca-model.c
+++ b/plugins/thunar-uca/thunar-uca-model.c
@@ -1038,22 +1038,8 @@ thunar_uca_model_match (ThunarUcaModel *uca_model,
   files = g_new (ThunarUcaFile, n_files);
   for (lp = file_infos, n = 0; lp != NULL; lp = lp->next, ++n)
     {
-      location = thunarx_file_info_get_location (lp->data);
-
-      if (!g_file_has_uri_scheme (location, "file"))
-        {
-          /* cannot handle non-local files */
-          g_object_unref (location);
-          g_free (files);
-          return NULL;
-        }
-
-      g_object_unref (location);
-
       info = thunarx_file_info_get_file_info (lp->data);
-
       mime_type = g_file_info_get_content_type (info);
-
       files[n].name = g_file_info_get_name (info);
       files[n].types = types_from_mime_type (mime_type);
 
